bundles/org.openhab.binding.mercedesme/src/main/java/org/openhab/binding/mercedesme/internal/server/CallbackServer.java

```diff
@@ -0,0 +1,190 @@
+import java.util.Optional;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.server.Connector;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.ServerConnector;
+import org.eclipse.jetty.servlet.ServletHandler;
+import org.openhab.binding.mercedesme.internal.AccountConfiguration;
+import org.openhab.binding.mercedesme.internal.Constants;
+import org.openhab.core.auth.client.oauth2.AccessTokenRefreshListener;
+import org.openhab.core.auth.client.oauth2.AccessTokenResponse;
+import org.openhab.core.auth.client.oauth2.OAuthClientService;
+import org.openhab.core.auth.client.oauth2.OAuthException;
+import org.openhab.core.auth.client.oauth2.OAuthFactory;
+import org.openhab.core.auth.client.oauth2.OAuthResponseException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link CallbackServer} class defines an HTTP Server for authentication callbacks
+ *
+ * @author Bernd Weymann - Initial contribution
+ */
+@NonNullByDefault
+public class CallbackServer {
+    private static final Logger LOGGER = LoggerFactory.getLogger(CallbackServer.class);
+    private static final Map<Integer, OAuthClientService> AUTH_MAP = new HashMap<Integer, OAuthClientService>();
+    private static final Map<Integer, CallbackServer> SERVER_MAP = new HashMap<Integer, CallbackServer>();
+
+    private Optional<Server> server = Optional.empty();
+    private Optional<AccessTokenResponse> token = Optional.empty();
+    private AccessTokenRefreshListener listener;
+    private AccountConfiguration config;
+    private OAuthClientService oacs;
+    private String callbackUrl;
+
+    public CallbackServer(AccessTokenRefreshListener l, HttpClient hc, OAuthFactory oAuthFactory,
+            AccountConfiguration config, String callbackUrl) {
+        oacs = oAuthFactory.createOAuthClientService(Constants.OAUTH_CLIENT_NAME + config.callbackPort,
+                Constants.MB_TOKEN_URL, Constants.MB_AUTH_URL, config.clientId, config.clientSecret, config.getScope(),
+                false);
+        AUTH_MAP.put(Integer.valueOf(config.callbackPort), oacs);
+        SERVER_MAP.put(Integer.valueOf(config.callbackPort), this);
+        listener = l;
+        this.config = config;
+        this.callbackUrl = callbackUrl;
+    }
+
+    public String getAuthorizationUrl() {
+        try {
+            return oacs.getAuthorizationUrl(callbackUrl, null, null);
+        } catch (OAuthException e) {
+            LOGGER.error("Error creating Authorization URL {}", e.getMessage());
+            return Constants.EMPTY;
+        }
+    }
+
+    public String getScope() {
+        return config.getScope();
+    }
+
+    public void start() {
+        LOGGER.debug("Start Callback Server for port {}", config.callbackPort);
+        if (!server.isEmpty()) {
+            LOGGER.debug("Callback server for port {} already started", config.callbackPort);
+            return;
```

Comment:
It seems the error handling here is quite complex. Perhaps it would be simpler to throw a dedicated exception here with this error message (or even user-oriented i18n string), so the caller could handle this appropriately and possibly log it (if needed)? The current flow is hard to understand, I'm not even sure if fully correct:
- `AccountHandler.initialize()` calls `setToken()` in this class if it can be found.
- Otherwise it logs "Token not found in storage - manual authorization needed"
- After knowing that token was not found, it calls this method.
- This method will then log "Callback server for port {} already started".
- `initialize()` will then call `getToken()`.
- If token is empty, it will now set status to OFFLINE with description "Manual Authorization needed at {0}".

Reply:
I see, responsibilities are a bit mixed up. I took a deeper look and recognized  `OAuthClientService`  is doing much more than expected from the interface description. It holds and persists the last token and also refreshing takes place if time is expired. So I did some rework:

-  `CallbackServer` handles restoring, storing and refreshing with the help of  `OAuthClientService`
-  `AccountHandler` is only responsible to switch the status based on the token response

So I was able to delete some unecessary code and responsibiliteis shall look better now

Reply:
Resolving this comment after the provided refactoring.